src/prefix-color-selector.ts

```diff
@@ -0,0 +1,77 @@
+import chalk from 'chalk';
+
+export class PrefixColorSelector {
+    lastColor: string;
+    autoColors: string[];
+
+    get ACCEPTABLE_CONSOLE_COLORS() {
+        // Colors picked randomly, can be amended if required
+        return (
+            [
+                chalk.cyan,
+                chalk.yellow,
+                chalk.magenta,
+                chalk.grey,
+                chalk.bgBlueBright,
+                chalk.bgMagenta,
+                chalk.magentaBright,
+                chalk.bgBlack,
+                chalk.bgWhite,
+                chalk.bgCyan,
+                chalk.bgGreen,
+                chalk.bgYellow,
+                chalk.bgRed,
+                chalk.bgGreenBright,
+                chalk.bgGrey,
+                chalk.blueBright,
+            ]
+                // Filter out duplicates
+                .filter((chalkColor, index, arr) => {
+                    return arr.indexOf(chalkColor) === index;
+                })
+                .map(chalkColor => chalkColor.bold.toString())
+        );
+    }
+
+    constructor(private readonly prefixColors?: string[], private readonly color?: boolean) {}
+
+    getNextColor(index?: number) {
+        const cannotSelectColor = !this.prefixColors?.length && !this.color;
+        if (cannotSelectColor) {
+            return '';
+        }
+
+        const userDefinedColorForCurrentCommand =
+            this.prefixColors && typeof index === 'number' && this.prefixColors[index];
+
+        if (!this.color) {
+            // Use documented behaviour of repeating last color
+            // when specifying more commands than colors
+            this.lastColor = userDefinedColorForCurrentCommand || this.lastColor;
+            return this.lastColor;
+        }
+
+        // User preference takes priority if defined
+        if (userDefinedColorForCurrentCommand) {
+            this.lastColor = userDefinedColorForCurrentCommand;
+            return userDefinedColorForCurrentCommand;
+        }
+
+        // Auto selection requested and no user preference defined, select next auto color
+        if (!this.autoColors || !this.autoColors.length) {
+            this.refillAutoColors();
+        }
+
+        // Prevent consecutive colors from being the same
+        // (i.e. when transitioning from user colours to auto colours)
+        const nextColor = this.autoColors.shift();
```

Comment:
This method has a mix of stateful and stateless behaviour; you can specify an index that is within the bounds of `prefixColors`, and it'll always return that (stateless/idempotent).

But otherwise, the result depends on how many times `getNextColor` was called before (stateful/random), so `index` becomes meaningless.

I'd like to get rid of this ambiguity, so here are some suggestions:
1. Remove the `index` argument. Let `getNextColor` always return a "random" value depending on how many times it was called before.
2. Iterate up to `index` times generating the colour for each index in a predictable way and return it. E.g. calling `getNextColor(10); getNextColor(10)` always returns the same value. This can make the method `O(N)`, so some optimisation is required.

Reply:
The result also depends on the user colours provided to prevent repetition, which affects the order of the colours. It should be repeatable for the same initial conditions, but results will vary with different user inputs, which I think is acceptable in this case but please let me know if you feel strongly about it.

I will look into this to understand the code again, however looking at it after a while I don't think it's a good implementation. What I was trying to do is have the getNext colour method act as an infinite stream of colours but it looks odd giving it an index. 

What do you think about using a generator function for this? Would you allow that in the code base?

Reply:
implementation changed to use a generator, awaiting review